helm/config_test.go

```diff
@@ -0,0 +1,813 @@
+# comment 1
+{{- if block 1 }}
+# comment 2
+{{- if block 2 }}
+Object:
+# comment 3
+{{- if block 3 }}
+- # comment 4
+  {{- if block 4 }}
+  Foo: foo
+  {{- end }}
+  # comment 5
+  {{- if block 5 }}
+  Bar: bar
+  {{- end }}
+  # comment 6
+  {{- if block 6 }}
+  Baz: baz
+  {{- end }}
+{{- end }}
+{{- end }}
+{{- end }}
+`)
+}
+
+func TestHelmMultiLineComment(t *testing.T) {
+	root := NewObject()
+	root.Add("Scalar", NewScalar("42", Comment("Many\n\nlines")))
+
+	equal(t, root, `---
+# Many
+#
+# lines
+Scalar: 42
+`)
+
+	// list of list
+	list1 := NewList()
+	list1.Add(NewScalar("42", Comment("Many\n\nlines")))
+
+	list2 := NewList()
+	list2.Add(list1)
+	list2.Add(NewScalar("foo"))
+
+	root = NewObject()
+	root.Add("List", list2)
+
+	equal(t, root, `---
+List:
+- # Many
+  #
+  # lines
+  - 42
+- foo
+`)
+}
+
+func TestHelmWrapLongComments(t *testing.T) {
+	root := NewObject()
+	obj := NewObject()
+	word := "1"
+	for i := len(word) + 1; i < 7; i++ {
+		word += strconv.Itoa(i)
+		root.Add(fmt.Sprintf("Key%d", i), NewScalar("~", Comment(strings.Repeat(word+" ", 10))))
+		if i < 5 {
+			obj.Add(fmt.Sprintf("Key%d", i), NewScalar("~", Comment(strings.Repeat(word+" ", 5))))
+		}
+	}
+
+	obj.Add("Very", NewScalar("Long", Comment(strings.Repeat(strings.Repeat("x", 50)+" ", 2))))
+	root.Add("Very", NewScalar("Long", Comment(strings.Repeat(strings.Repeat("x", 50)+" ", 2))))
+	root.Add("Nested", obj)
+
+	expect := `---
+# 12 12 12 12 12 12 12
+# 12 12 12
+Key2: ~
+# 123 123 123 123 123
+# 123 123 123 123 123
+Key3: ~
+# 1234 1234 1234 1234
+# 1234 1234 1234 1234
+# 1234 1234
+Key4: ~
+# 12345 12345 12345
+# 12345 12345 12345
+# 12345 12345 12345
+# 12345
+Key5: ~
+# 123456 123456 123456
+# 123456 123456 123456
+# 123456 123456 123456
+# 123456
+Key6: ~
+# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+Very: Long
+Nested:
+          # 12 12 12 12
+          # 12
+          Key2: ~
+          # 123 123 123
+          # 123 123
+          Key3: ~
+          # 1234 1234
+          # 1234 1234
+          # 1234
+          Key4: ~
+          # xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+          # xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+          Very: Long
+`
+	equal(t, root, expect, Indent(10), Wrap(24))
+}
+
+func TestHelmIndent(t *testing.T) {
+	obj1 := NewObject()
+	obj1.Add("Foo", NewScalar("Bar", Comment("Baz")))
+
+	list1 := NewList()
+	list1.Add(obj1)
+	list1.Add(NewScalar("1"))
+
+	list2 := NewList()
+	list2.Add(NewScalar("abc"))
+	list2.Add(NewScalar("xyz"))
+
+	list1.Add(list2)
+	list1.Add(NewScalar("2"))
+	list1.Add(NewScalar("3"))
+
+	obj2 := NewObject()
+	obj2.Add("List", list1)
+
+	obj3 := NewObject()
+	obj3.Add("Foo", NewScalar("1"))
+	obj3.Add("Bar", NewScalar("2"))
+
+	obj2.Add("Meta", obj3)
+
+	root := NewObject()
+	root.Add("Object", obj2)
+
+	expect := `---
+Object:
+    List:
+      - # Baz
+        Foo: Bar
+      - 1
+      -   - abc
+          - xyz
```

Comment:
Not sure which looks better:
```yaml
List:
-   foo
-   bar
```
```yaml
List:
  - foo
  - bar
```

Reply:
I don't know either, but for some reason I thought I preferred the latter. Now I'm not sure anymore. Looking at more complex documents and large indents may help to make up your mind (or not).

```
Mapping:
        Foo: abc
        Bar: xyz
        List:
                Foo: abc
                Bar: xyz
CurrentList:
      - abc
      - xyz
      -       - abc
              - xyz
Alternative:
-       abc
-       xyz
-       -       abc
        -       xyz
```

I slightly prefer keeping the "dash" close to the value (seems easier to read because you don't have to visually align over a lot of whitespace), but have no big objections to changing this. Realistically we are probably keeping the `indent` at 2, in which case there is no difference to begin with.